Regression: Using TestNG via Maven breaks when optional Guice dependency is unavailable

### TestNG Version
7.4.0-SNAPSHOT (built from bdf1c95501d9b6afb8a6e1a038dbdfbbf7acc378)

### Expected behavior
TestNG should be able to execute simple tests without requiring Guice to be present on the classpath.

### Actual behavior
TestNG fails to start when Guice is not present on the classpath, due to ee6cbabdbf3b5c3eec7685007e4d2d97136308d1.

### Is the issue reproductible on runner?

- [ ] Shell
- [x] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
`pom.xml`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>testng-2403</artifactId>
    <version>1.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>org.testng</groupId>
            <artifactId>testng</artifactId>
            <version>7.4.0-SNAPSHOT</version>
            <scope>test</scope>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>2.19</version>
            </plugin>
        </plugins>
    </build>

</project>
```

`src/test/java/example/TestClass.java` (this doesn't really matter, but one test class must exist for TestNG to execute):
```java
package example;

import org.testng.annotations.Test;

public class TestClass {

    @Test
    public void test() {
        // this should pass
    }
}
```

Actual output:
```
$ mvn clean test
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------------< org.example:testng-2403 >-----------------------
[INFO] Building testng-2403 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ testng-2403 ---
[INFO] Deleting /Users/morgan/src/testng-2403/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ testng-2403 ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] Copying 0 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ testng-2403 ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ testng-2403 ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] skip non existing resourceDirectory /Users/morgan/src/testng-2403/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ testng-2403 ---
[INFO] Changes detected - recompiling the module!
[WARNING] File encoding has not been set, using platform encoding UTF-8, i.e. build is platform dependent!
[INFO] Compiling 1 source file to /Users/morgan/src/testng-2403/target/test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.19:test (default-test) @ testng-2403 ---

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running example.TestClass

Results :

Tests run: 0, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.105 s
[INFO] Finished at: 2020-11-04T13:19:12-06:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19:test (default-test) on project testng-2403: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.19:test failed: There was an error in the forked process
[ERROR] java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: com/google/inject/Module
[ERROR] 	at org.testng.GuiceHelper.<clinit>(GuiceHelper.java:28)
[ERROR] 	at org.testng.TestRunner.<init>(TestRunner.java:87)
[ERROR] 	at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:593)
[ERROR] 	at org.testng.SuiteRunner.init(SuiteRunner.java:173)
[ERROR] 	at org.testng.SuiteRunner.<init>(SuiteRunner.java:107)
[ERROR] 	at org.testng.TestNG.createSuiteRunner(TestNG.java:1300)
[ERROR] 	at org.testng.TestNG.createSuiteRunners(TestNG.java:1276)
[ERROR] 	at org.testng.TestNG.runSuitesLocally(TestNG.java:1125)
[ERROR] 	at org.testng.TestNG.runSuites(TestNG.java:1063)
[ERROR] 	at org.testng.TestNG.run(TestNG.java:1031)
[ERROR] 	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:132)
[ERROR] 	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:134)
[ERROR] 	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:118)
[ERROR] 	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
[ERROR] 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:286)
[ERROR] 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:240)
[ERROR] 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
[ERROR] Caused by: java.lang.NoClassDefFoundError: com/google/inject/Module
[ERROR] 	... 17 more
[ERROR] Caused by: java.lang.ClassNotFoundException: com.google.inject.Module
[ERROR] 	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
[ERROR] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
[ERROR] 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
[ERROR] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
[ERROR] 	... 17 more
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

### Previous Behavior

If you change `7.4.0-SNAPSHOT` to `7.3.0` (or an earlier version), and you add an exclusion for Guice (to eliminate it from the classpath), then it passes.

`pom.xml`:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>testng-2403</artifactId>
    <version>1.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>org.testng</groupId>
            <artifactId>testng</artifactId>
            <version>7.3.0</version>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>org.testng</groupId>
                    <artifactId>testng</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>2.19</version>
            </plugin>
        </plugins>
    </build>

</project>
```

Output:
```
$ mvn clean test
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------------< org.example:testng-2403 >-----------------------
[INFO] Building testng-2403 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ testng-2403 ---
[INFO] Deleting /Users/morgan/src/testng-2403/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ testng-2403 ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] Copying 0 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ testng-2403 ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ testng-2403 ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] skip non existing resourceDirectory /Users/morgan/src/testng-2403/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ testng-2403 ---
[INFO] Changes detected - recompiling the module!
[WARNING] File encoding has not been set, using platform encoding UTF-8, i.e. build is platform dependent!
[INFO] Compiling 1 source file to /Users/morgan/src/testng-2403/target/test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.19:test (default-test) @ testng-2403 ---

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running example.TestClass
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.381 sec - in example.TestClass

Results :

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.587 s
[INFO] Finished at: 2020-11-04T13:23:39-06:00
[INFO] ------------------------------------------------------------------------
```

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2405: Regression: Using TestNG via Maven breaks when optional Guice dependency is unavailable (Krishnan Mahadevan)
 Fixed: GITHUB-2427: Guice module (suite parent-module and test module) configure() method is called multiple times (Jacek Centkowski)
 Fixed: GITHUB-2419: TestNG JUnit reports are not valid if system output contains XML tags (Lorenzo Orsatti)
 Fixed: GITHUB-188: suite parallel="methods" does not work when there are multiple <test> tags in the testng.xml (Krishnan Mahadevan)

--- a/src/main/java/org/testng/TestRunner.java
+++ b/src/main/java/org/testng/TestRunner.java
@@ -16,6 +16,7 @@ import java.util.concurrent.BlockingQueue;
 
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
+import org.testng.annotations.Guice;
 import org.testng.collections.ListMultiMap;
 import org.testng.collections.Lists;
 import org.testng.collections.Maps;
@@ -50,6 +51,7 @@ import org.testng.internal.TestNGClassFinder;
 import org.testng.internal.TestNGMethodFinder;
 import org.testng.internal.Utils;
 import org.testng.internal.XmlMethodSelector;
+import org.testng.internal.annotations.AnnotationHelper;
 import org.testng.internal.annotations.IAnnotationFinder;
 import org.testng.thread.ITestNGThreadPoolExecutor;
 import org.testng.thread.IThreadWorkerFactory;
@@ -84,7 +86,7 @@ public class TestRunner
   private String m_testName;
   private IInjectorFactory m_injectorFactory;
 
-  private final GuiceHelper guiceHelper = new GuiceHelper(this);
+  private GuiceHelper guiceHelper;
 
   private List<XmlClass> m_testClassesFromXml = null;
 
@@ -1285,6 +1287,12 @@ public class TestRunner
 
   @Override
   public Injector getInjector(IClass iClass) {
+    if (hasNoGuiceAnnotations(iClass)) {
+      return null;
+    }
+    if (guiceHelper == null) {
+      guiceHelper = new GuiceHelper(this);
+    }
     return guiceHelper.getInjector(iClass, this.m_injectorFactory);
   }
 
@@ -1292,4 +1300,9 @@ public class TestRunner
   public void addInjector(List<Module> moduleInstances, Injector injector) {
     m_injectors.put(moduleInstances, injector);
   }
+
+  private static boolean hasNoGuiceAnnotations(IClass iClass) {
+    return AnnotationHelper.findAnnotationSuperClasses(Guice.class, iClass.getRealClass()) == null;
+  }
+
 }

--- a/src/test/java/test/thread/issue188/Issue188TestSample.java
+++ b/src/test/java/test/thread/issue188/Issue188TestSample.java
@@ -1,8 +1,10 @@
 package test.thread.issue188;
 
 import java.util.Map;
+import java.util.Random;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
 import org.testng.ITestResult;
 import org.testng.Reporter;
 import org.testng.annotations.BeforeMethod;
@@ -10,6 +12,7 @@ import org.testng.annotations.Test;
 
 public class Issue188TestSample {
   public static final Map<Long, Set<String>> timestamps = new ConcurrentHashMap<>();
+  private static final Random random = new Random();
 
   @BeforeMethod
   public void logTime(ITestResult itr) {
@@ -19,10 +22,20 @@ public class Issue188TestSample {
 
   @Test
   public void sampleTest() {
+    sleepSilently();
   }
 
   @Test
   public void anotherSampleTest() {
+    sleepSilently();
+  }
+
+  private void sleepSilently() {
+    try {
+      TimeUnit.MILLISECONDS.sleep(500 * random.nextInt(10));
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
   }
 
 }

--- a/src/test/java/test/thread/issue188/IssueTest.java
+++ b/src/test/java/test/thread/issue188/IssueTest.java
@@ -1,7 +1,10 @@
 package test.thread.issue188;
 
 import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
 import java.util.Set;
+import java.util.stream.Collectors;
 import org.assertj.core.api.Assertions;
 import org.testng.TestNG;
 import org.testng.annotations.Test;
@@ -32,11 +35,22 @@ public class IssueTest extends SimpleBaseTest {
           .withFailMessage("Since all tests were started simultaneously,test method count should have been 6")
           .hasSize(6);
       } else {
-        long answer = Issue188TestSample.timestamps.keySet().stream()
-            .reduce(0L, (x, y) -> Math.abs(x - y));
-        Assertions.assertThat(answer)
-            .withFailMessage("test methods should have started within a lag of max 20 ms")
-            .isLessThanOrEqualTo(20);
+        List<Long> keyset = Issue188TestSample.timestamps.keySet().stream()
+            .sorted().collect(Collectors.toList());
+        String allTimeStamps = keyset.stream()
+            .map(Objects::toString)
+            .collect(Collectors.joining(","));
+        long prev = keyset.get(0);
+        for (int i = 1; i < keyset.size(); i++) {
+          long current = keyset.get(i);
+          long diff = current - prev;
+          Assertions.assertThat(diff)
+              .withFailMessage(
+                  "Test methods should have started within a lag of max 40 ms but it was "
+                      + diff + " ms [" + allTimeStamps + "]")
+              .isLessThanOrEqualTo(40);
+          prev = current;
+        }
       }
     } finally {
       System.setProperty(RuntimeBehavior.STRICTLY_HONOUR_PARALLEL_MODE, "false");

